### PR TITLE
Multiclassdirichlet give uncertainty

### PR DESCRIPTION
**Pull request by KnathanM**
## Description
This PR adds an uncertainty predictor for multiclass dirichlet.

`BinaryDirichletFFN` returns an `n x t x s` tensor where `n` is the batch size, `t` is the number of tasks and `s` is the number of targets, in this case predictions and uncertainties. This makes it similar to `MveFFN`. Dirichlet is unusual because it is a classification method, but doesn't use the class probabilities as the uncertainties. Rather it uses the amount of "evidence" it has for each class to calculate the uncertainty. 

My understanding is that the dirichlet method implemented in Chemprop predicts an amount of "evidence" (alpha) > 1 for each class. The probability of each class is its amount of evidence divided by the total evidence across all classes, and the uncertainty is the number of classes divided by the total amount of evidence. If we have no evidence for any class (i.e. alpha for each class is 1), then each class will have a 1/num_classes probability and the uncertainty will be num_classes/num_classes = 1. 

There is no difference between binary dirichlet and multiclass dirichlet. This is reflected in the fact that now that `BinaryDirichletFFN.train_step()` returns an `n x t x 2` tensor, we don't need a separate loss function for binary vs multiclass. 

We still have to keep `BinaryDirichletFFN` separate from `MulticlassDirichletFFN` though because the rest of Chemprop is set up to differentiate between binary prediction and multiclass prediction (different metrics, etc.).

The main problem this PR solves is how should we give the uncertainty prediction from multiclass dirichlet to users. For binary, we can return it in the 2nd dimension (zero indexed) like we do for MVE. But multiclass already uses the 2nd dimension for the probability of each class. 

We could treat the uncertainty as another class and return an `n x t x (c+1)` tensor, but that seems a bit over engineered. 

Instead we don't have `MulticlassDirichletFFN` return uncertainty and only return the class probabilities. To get the uncertainty we hack `MulticlassDirichletFFN` a bit in the uncertainty module `MulticlassDirichletPredictor`. We replace the forward method with one that gives both the class predictions and the uncertainties . Then we give them to the user as a tuple, which fits with the return type of the other uncertainty predictors. I think the uncertainty predictor for classification `ClassPredictor` will return the class probabilities as both the predictions and as the uncertainties for example.

### CODE REVIEW

**Comment by shihchengli** [comment]
Yes, `ClassPredictor` will treat the probability as uncertainty as well. I am wondering if we should implement this method, as it essentially provides no additional information.

**Comment by shihchengli** [comment]
After a second thought, I think it doesn't hurt to add this, as it can make the implementation of uncertainty in the CLI more straightforward. I have added the code in #1072. Feel free to let me know if you have any thoughts.

File: chemprop/uncertainty/predictor.py [hunk 1]
```diff
@@ -291,10 +310,56 @@ def __call__(
...
+        preds = []
+        uncs = []
+        for model in models:
+            self._setup_model(model)
+            output = torch.concat(trainer.predict(model, dataloader), 0)
+            self._restore_model(model)
+            preds.append(output[..., :-1])
+            uncs.append(output[..., -1])
+            preds = torch.stack(preds, 0)
+            uncs = torch.stack(uncs, 0).unsqueeze(-1).repeat(1, 1, 1, preds.shape[-1])
```
Comment (shihchengli) [inline chemprop/uncertainty/predictor.py:344]:
```suggestion
        preds = torch.stack(preds, 0)
        uncs = torch.stack(uncs, 0)
```
Should we change the docstring to reflect the shape of multiclass classification as `m' n x t` instead of `m' n x t x c`?

File: tests/unit/uncertainty/test_predictors.py [hunk 2]
```diff
@@ -70,3 +73,35 @@ def test_EnsemblePredictor_wrong_n_models():
...
+def test_MulticlassDirichletPredictor(data_dir, dataloader, trainer):
+    model = MPNN.load_from_file(data_dir / "example_model_v2_multiclass_dirichlet_mol.pt")
+    predictor = MulticlassDirichletPredictor()
+    preds, uncs = predictor(dataloader, [model], trainer)
+
+    torch.testing.assert_close(
+        preds, torch.tensor([[[[0.906426, 0.046787, 0.046787]], [[0.925395, 0.037303, 0.037303]]]])
+    )
+    torch.testing.assert_close(
+        uncs, torch.tensor([[[[0.140361, 0.140361, 0.140361]], [[0.111908, 0.111908, 0.111908]]]])
```
Comment (shihchengli) [inline tests/unit/uncertainty/test_predictors.py:106]:
```suggestion
        uncs, torch.tensor([[[0.140361], [0.111908]]])
```
